semgrep-core/synthesizing/Pattern_from_Code.ml

```diff
@@ -28,25 +28,146 @@ open AST_generic
 type named_variants =
   (string * Pattern.t) list
 
+type global_state = { count : int; mapping : (expr * expr) list;
```

Comment:
We usually calls that kind of thing an "environment". It's not really a state because
you don't modify it in place (it's not a global; it's something you pass around).